### PR TITLE
Cap 0035 set trustline flags

### PR DESCRIPTION
**Pull request by sisuresh**
# Description

PR for the new `setTrustLineFlagsOp` specified in https://github.com/stellar/stellar-protocol/blob/master/core/cap-0035.md.

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/SetTrustLineFlagsOpFrame.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            expectedFlagValue &= ~*mSetTrustLineFlags.clearFlags;
+        }
+
+        if (mSetTrustLineFlags.setFlags)
+        {
+            expectedFlagValue |= *mSetTrustLineFlags.setFlags;
+        }
+
+        shouldRemoveOffers =
+            isAuthorizedToMaintainLiabilities(trust) && expectedFlagValue == 0;
```
Comment (jonjove) [inline src/transactions/SetTrustLineFlagsOpFrame.cpp:101]: This doesn't seem right to me. What if `expectedFlagValue == TRUSTLINE_CLAWBACK_ENABLED_FLAG`?

File: src/transactions/SetTrustLineFlagsOpFrame.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    if (!trustLineFlagAuthIsValid(expectedFlagValue))
```
Comment (jonjove) [inline src/transactions/SetTrustLineFlagsOpFrame.cpp:104]: Thoughts on moving this to before where `shouldRemoveOffers` is defined? My main justifications are:
- it minimizes the interleaving of getting the new flags with calculating `shouldRemoveOffers`
- why do the (admittedly minimal) work to compute `shouldRemoveOffers` if you know you are going to fail

I'm flexible on this if you disagree.

File: src/transactions/SponsorshipUtils.cpp [hunk 3]
```diff
@@ -753,6 +753,13 @@ removeEntryWithPossibleSponsorship(AbstractLedgerTxn& ltx,
 
         if (acc.current().data.account().accountID == *le.ext.v1().sponsoringID)
         {
+            // the "owner" of a subentry cannot also be the sponsor of that
```
Comment (jonjove) [inline src/transactions/SponsorshipUtils.cpp:756]: What does this have to do with this PR?

**Review by jonjove — COMMENTED** [review]
Looks mostly good. Several minor comments plus one thing that I'm concerned about. I haven't reviewed any of the tests yet, and I'll want to take a second look over everything when I do that.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp (hunk 1 above)
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp:101]: Wow good catch. I'll add a test for this. Good reminder that the CAP-0035 changes still need to be tested thoroughly.

File: src/transactions/SponsorshipUtils.cpp (hunk 3 above)
Comment (sisuresh) [reply src/transactions/SponsorshipUtils.cpp:756]: It doesn't. I wanted to add this comment and thought I'd squeeze it into this PR instead of creating a new one. I can open a separate PR instead.

File: src/transactions/TransactionUtils.cpp [hunk 4]
```diff
@@ -994,6 +995,33 @@ getAsset(AccountID const& issuer, AssetCode const& assetCode)
     return asset;
 }
 
+void
+removeOffersByAccountAndAsset(AbstractLedgerTxn& ltx,
+                              LedgerTxnHeader const& header,
+                              AccountID const& account, Asset const& asset)
+{
+    auto offers = ltx.loadOffersByAccountAndAsset(account, asset);
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Done

File: src/transactions/SetTrustLineFlagsOpFrame.cpp [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        auto header = ltxSource.loadHeader();
+        auto sourceAccountEntry = loadSourceAccount(ltxSource, header);
+        auto const& sourceAccount = sourceAccountEntry.current().data.account();
+
+        const uint32_t authFlags =
+            AUTHORIZED_FLAG | AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG;
+
+        bool authNotRevocable = !(sourceAccount.flags & AUTH_REVOCABLE_FLAG);
+
+        // This condition covers all three revoke scenarios (AUTHORIZED_FLAG ->
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp]: I simplified the comment.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+{
+    ZoneNamedN(applyZone, "SetTrustLineFlagsOp apply", true);
+
+    {
+        LedgerTxn ltxSource(ltx); // ltxSource will be rolled back
+        auto header = ltxSource.loadHeader();
+        auto sourceAccountEntry = loadSourceAccount(ltxSource, header);
+        auto const& sourceAccount = sourceAccountEntry.current().data.account();
+
+        const uint32_t authFlags =
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp]: I actually removed this and added a `TRUSTLINE_AUTH_FLAGS` variable in `TransactionUtils.h` since this flag combo is used in a few places.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+{
+    return protocolVersion >= 16;
+}
+
+bool
+SetTrustLineFlagsOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    ZoneNamedN(applyZone, "SetTrustLineFlagsOp apply", true);
+
+    {
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp:44]: Done. Good idea!

File: src/transactions/SetTrustLineFlagsOpFrame.cpp (hunk 1 above)
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp]: Fixed and added a testcase.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp (hunk 2 above)
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp]: Yeah I think moving it makes sense. Done.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp:44]: This made me realize something I need to address - I haven't been consistent with the flag checks (`flags & AUTH_REVOCABLE_FLAG` vs `(flags & AUTH_REVOCABLE_FLAG) != 0)`). I'll fix this.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    // In all three scenarios above, clearingAnyAuth == true and
+    // settingAuthorized == false
+
+    bool clearingAnyAuth =
+        mSetTrustLineFlags.clearFlags &&
+        (*mSetTrustLineFlags.clearFlags & TRUSTLINE_AUTH_FLAGS) != 0;
+
+    bool settingAuthorized =
+        mSetTrustLineFlags.setFlags &&
+        (*mSetTrustLineFlags.setFlags & AUTHORIZED_FLAG) != 0;
```
Comment (MonsieurNicolas) [inline src/transactions/SetTrustLineFlagsOpFrame.cpp:196]: Note that the CAP is under-specified when attempting to transition from 0 to one of the authorized states when the issuer does not have `AUTH_REVOCABLE_FLAG`.

This returns false (and causes `SET_TRUST_LINE_FLAGS_CANT_REVOKE`) when trying to transition from `0 -> AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG`).

Also, this returns true for `0 -> AUTHORIZED_FLAG` - yet `AllowTrustOp` fails in this situation with `ALLOW_TRUST_TRUST_NOT_REQUIRED` (in this PR). I think the CAP should be explicit on the fact we're aligning with CAP29 here.

In any case you want to to make sure that we have `iif` : `(revoke scenario) iif (clearingAnyAuth == true and settingAuthorized == false)`.

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]

**Comment by MonsieurNicolas** [comment]
I missed this in my review: the `LedgerEntry` invariant does not verify that we never re-enable the `TRUSTLINE_CLAWBACK_ENABLED_FLAG` flag

File: src/xdr/Stellar-transaction.x [hunk 9]
```diff
@@ -378,6 +379,24 @@ struct ClawbackOp
...
+   Threshold: low
+
+   Result: SetTrustLineFlagsResult
+*/
+struct SetTrustLineFlagsOp
+{
+    AccountID trustor;
+    Asset asset;
+
+    uint32* clearFlags; // which flags to clear
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/xdr/Stellar-transaction.x:396]: This was done to match the existing `SetOptionsOp` parameters.

[the remark this answers is not in the document]
Comment (jonjove) [reply src/xdr/Stellar-transaction.x:396]: @MonsieurNicolas does make an interesting observation here. Just because `SetOptionsOp` has a slightly strange interface doesn't mean that we have to propagate that to `SetTrustLineFlagsOp`. I know @leighmcculloch has a lot of thoughts about interface consistency, so I'm interested to hear his take on this.

[the remark this answers is not in the document]
Comment (leighmcculloch) [reply src/xdr/Stellar-transaction.x:396]: I don't think this has any meaningful impact one way or the other. Looks like you could drop the pointer to me.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp (hunk 8 above)
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp:196]: 1. For your point about `0 -> AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG` - This function returns true, and there's a test case for this under the section `upgrade auth when not revocable`. 
2. I agree the CAP should be updated to mention that we're aligning with CAP29. I'll take care of this.
3. I believe the condition I have here is equivalent to the condition in your comment. Am I missing something?

File: src/xdr/Stellar-transaction.x (hunk 9 above)
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/xdr/Stellar-transaction.x:396]: Yeah I don't feel strongly about either approach. If any of you do think we should drop the pointer, I'm fine with that. Let me know and I'll update this PR and the CAP.

**Comment by sisuresh** [comment]
> I missed this in my review: the `LedgerEntry` invariant does not verify that we never re-enable the `TRUSTLINE_CLAWBACK_ENABLED_FLAG` flag

Good catch. I updated `LedgerEntryIsValid`.

File: src/transactions/TransactionUtils.cpp [hunk 10]
```diff
@@ -863,9 +863,8 @@ trustLineFlagIsValid(uint32_t flag, uint32_t ledgerVersion)
     }
     else
     {
-        uint32_t invalidAuthCombo =
-            AUTHORIZED_FLAG | AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG;
-        if ((flag & invalidAuthCombo) == invalidAuthCombo)
+        // multiple auth flags can't be set
+        if ((flag & TRUSTLINE_AUTH_FLAGS) == TRUSTLINE_AUTH_FLAGS)
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: I added the `static_assert`.

File: src/transactions/TransactionUtils.cpp [hunk 11]
```diff
@@ -856,27 +856,37 @@ releaseLiabilities(AbstractLedgerTxn& ltx, LedgerTxnHeader const& header,
...
+{
+    return trustLineFlagMaskCheckIsValid(flag, ledgerVersion) &&
+           (ledgerVersion < 13 || trustLineFlagAuthIsValid(flag));
+}
+
+bool
+trustLineFlagAuthIsValid(uint32_t flag)
+{
+    // multiple auth flags can't be set
+    if ((flag & TRUSTLINE_AUTH_FLAGS) == TRUSTLINE_AUTH_FLAGS)
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:880]: I added the `static_assert`.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp (hunk 8 above)
Comment (MonsieurNicolas) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp:196]: Oh I see - I got confused by the comment the logic itself is fine.

What this is really doing is enforce that either auth flags are not being unset or if they are, the `AUTHORIZED_FLAG` is set so that we end up with the most restrictive auth.

The comment talks about it backwards (revoke scenarios) instead of talking about which case is valid.

**Review by jonjove — COMMENTED** [review]
This is very close now. I think there might be an opportunity to make some test improvements.

File: src/transactions/test/AllowTrustTests.cpp [hunk 12]
```diff
@@ -327,27 +375,62 @@ TEST_CASE("allow trust", "[tx][allowtrust]")
...
-            {
-                REQUIRE_THROWS_AS(gateway.allowTrust(idr, a1),
-                                  ex_ALLOW_TRUST_NO_TRUST_LINE);
-                REQUIRE_THROWS_AS(gateway.denyTrust(idr, a1),
-                                  ex_ALLOW_TRUST_CANT_REVOKE);
-            }
-            SECTION("set revocable flag")
-            {
-                gateway.setOptions(setFlags(AUTH_REVOCABLE_FLAG));
+        SECTION("allow trust without trustline")
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/test/AllowTrustTests.cpp:378]: I think the duplication in this commit is minimal, and is mainly due to the difference in exceptions thrown (`ex_ALLOW_TRUST_NO_TRUST_LINE` vs `ex_SET_TRUST_LINE_FLAGS_CANT_REVOKE`). Deduplicating all of those sections would make them harder to read and isn't worth the effort in my opinion. 

Do you have any ideas? `REQUIRE_THROWS_AS` takes the exception type name as an input so I'm not sure if there's a simple way to generalize those calls.

File: src/test/TxTests.h [hunk 13]
```diff
@@ -59,6 +59,16 @@ struct SetOptionsArguments
                                          SetOptionsArguments const& y);
 };
 
+struct SetTrustLineFlagsArguments
+{
+    uint32_t setFlags = 0;
+    uint32_t clearFlags = 0;
+
+    friend SetTrustLineFlagsArguments
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/test/TxTests.h:67]: The operator is defined as a member function, but is declared as a non member function in `TxTests.cpp`, so it needs to be defined as a `friend` (I modeled this after the existing `SetOptionsArguments`). I can change this by moving the operator out of the `SetTrustLineFlagsArguments` struct if you think that's worth doing.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    return true;
+}
+
+void
+SetTrustLineFlagsOpFrame::insertLedgerKeysToPrefetch(
+    UnorderedSet<LedgerKey>& keys) const
+{
+    if (mSetTrustLineFlags.asset.type() != ASSET_TYPE_NATIVE)
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp:145]: Yeah that's a good point. I'll change this to throw if the asset is native.

File: src/test/TxTests.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (jonjove) [reply src/test/TxTests.h:67]: Upon rereading, this comment was pretty unclear. What I meant was "This is a `struct` so you can access the data (which is `public` by default) from a non-friend non-member, so you should change this to be a non-friend non-member." `SetOptionsArguments` should have been done the same way.

File: src/transactions/SetTrustLineFlagsOpFrame.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp]: Fixed.

File: src/test/TxTests.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/test/TxTests.h]: Done

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]
looking good!

File: src/transactions/AllowTrustOpFrame.cpp [hunk 15]
```diff
@@ -59,7 +55,8 @@ bool
 AllowTrustOpFrame::doApply(AbstractLedgerTxn& ltx)
 {
     ZoneNamedN(applyZone, "AllowTrustOp apply", true);
-    if (ltx.loadHeader().current().ledgerVersion > 2)
+    auto ledgerVersion = ltx.loadHeader().current().ledgerVersion;
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/AllowTrustOpFrame.cpp:58]: That's an interesting idea that I hadn't considered. The point about the exact behavior is a good one. I didn't think `AllowTrustOpFrame` would need to change in the future though since it should never be able to modify any future flags, so there shouldn't be anything to copy/paste. I think If we find ourselves in that position in the future, we can implement your idea then.

[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/transactions/AllowTrustOpFrame.cpp:58]: yes that's fine - I am thinking more about some of the edge cases that come with auth revocation

File: src/transactions/test/AllowTrustTests.cpp [hunk 16]
```diff
@@ -10,463 +10,491 @@
...
 #include "test/TxTests.h"
 #include "test/test.h"
+#include "transactions/TransactionUtils.h"
 #include "util/Timer.h"
 
 using namespace stellar;
 using namespace stellar::txtest;
 
-TEST_CASE("authorized to maintain liabilities", "[tx][allowtrust]")
+namespace detail
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/test/AllowTrustTests.cpp:25]: Done as a fixup commit

**Review by jonjove — APPROVED** [review]
Looks good! We just need C++17 to get this merged (or a small fix at the end)

**Comment by sisuresh** [comment]
> Looks good! We just need C++17 to get this merged (or a small fix at the end)

I doubt we'll be on C++17 before next week so I'll work on the fix tomorrow.

**Comment by MonsieurNicolas** [comment]
r+ 0ed34d6b98dcabb26f65fd88de551abc9e4038ee
